Kernel/FileSystem/ProcFS/Inode.h

```diff
@@ -14,28 +14,90 @@
 
 namespace Kernel {
 
-class ProcFSInode : public Inode {
+class ProcFSExposedDirectory;
+class ProcFSExposedLink;
+class ProcFSInode final : public Inode {
     friend class ProcFS;
 
 public:
+    enum class Type {
+        GlobalLink,
+        GlobalDirectory,
+        FileDescriptionLink,
+        ThreadStack,
+        ProcessProperty,
+        ChildProcessLink,
+        ProcessDirectory,
+        ProcessSubdirectory,
```

Comment:
I'm a bit confused by the motivation here. In SysFS, we are adding more subclasses and types for each new piece of information. But in ProcFS, the desire is to move each piece of information into a single class with an enum to determine which object we've created? 

I guess my questions are:

- What does this buy us for future maintenance?
- Why do we not want the same optimization on SysFS?
- It looks like this will increate the object size of each node in procfs by the size of the optionals that it's not using (so sizeof(T) + sizeof(bool) for each). Is this a cost that we want to take on?

Reply:
- Well, I don't like how we currently have a clunky, over-complicated structure of the ProcFS, so I'd like to change it. It means a more static, simplified structure than now, to resemble what we had in the past.
- SysFS is meant to be expanded and will be expanding. On the other hand, ProcFS from my perspective will grow very slowly - maybe you will add another process data here or there, but the fundamental core of the ProcFS is not changed.
Also, it's really not the appropriate thing in my opinion to compare between the two filesystems - the SysFS code basically only expanded on its backend (i.e. internal objects), and the core inode code was not touched in about months by now and probably will not be touched again any time soon, because it works reliably and is structured really OK. On the other hand, the ProcFS backend code (its internal objects, not inodes) was not changed in months too. You could see that the SysFS inode code is very simple - only putting a distinction between the root directory inode, a symlink inode, a directory inode and a regular inode - which is far more sane than what we have in the ProcFS, to be honest. 
Now, due to the nature of the ProcFS - we simply don't allocate that much, or try to allocate only a minimal set of things (e.g. when the user actually accesses the filesystem), but on the SysFS, inodes are simply an interconnect to the internal objects, so they are not the bulky objects being around - we already allocated a lot more, but that's OK, because we actually accept it. The point is - both from the perspective of how an inode interfaces with internal kernel sources and the fact that the architectural point of view on both filesystems is not the same, I don't think we should apply this optimization on the SysFS.
- Hmm, that's a good question. I don't think the delta is that significant, to be honest, to worth scrapping the idea of this pull request. Also, this isn't the final word on the re-structure of the ProcFS I'd like to do, so we can probably find ways to decrease it even further.